### DISCUSSION TITLE
[@types/lodash] mapValues always returns an object but type may be inferred as an array

### DISCUSSION THREAD

**Discussion by skyboyer** [DefinitelyTyped/DefinitelyTyped#69439] [Issues with a @types package]
AFAIK, `_.mapValues` always return an object:

```
_.mapValues(undefined, () => 1); // {}
_.mapValues([1,2,3], () => true); // { 1: true, 2: true, 3: true }
_.mapValues('abc', () => '-'); // { 1: '-', 2: '-', 3: '-' }
```

But `@types/lodash` may result in an array-type inferred:
```
const a = mapValues([1,2,3], () => true) // { 1: true, 2: true, 3: true }
//    ^? const a: boolean[]
```

[Link to TS playground](https://www.typescriptlang.org/play/?ssl=3&ssc=9&pln=2&pc=1#code/JYWwDg9gTgLgBAbziAhmAaigNgVwKYDOcAvnAGZQQhwDkWEAJigQBY0DcAUAMYQB2BeCjgBeZGky5CACgDaARgA0AJkUBmALqK40gJSiAfHBhR8uzgHoLcG3AB6AfiA)

Is it a gap or do I miss some cases when it will be correct?

[a comment by an automated account is not shown]

**Comment by aj-r** [comment]
Interesting. I don't see any indication that `mapValues` was designed to be used for arrays, but it probably doesn't hurt to account for arrays in the typings.

The best solution is probably to add a new `mapValues` overload at the top of the overloads list. Something like:

```ts
mapValues<T, TResult>(array: T[], callback: (value: T) => TResult): NumericDictionary<TResult>;
```

**Reply by aj-r** [reply]
I don't have time to implement this but I'd be happy to review a PR.

**Comment by skyboyer** [comment]
created #69681
